### PR TITLE
Initial commit of go-proxy

### PR DESCRIPTION
**Pull request by vikramraman**
Completed:
- Listening for Wavefront format metrics, in-memory buffering and forwarding to Wavefront.
- Added agent metrics collection/reporting using go-metrics (https://github.com/rcrowley/go-metrics)

TODO:
- Need to expand on the agent metrics that is sent to Wavefront (include memory usage etc).
- Need to flesh out the proxy fetchConfig/checkin logic with the Wavefront servers.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,43 @@
+PREFIX := /usr/local
```
Comment (moribellamy) [inline Makefile:1]: if this needs to get more complicated, maybe lets look in to make alternatives. https://github.com/go-task/task and https://getgb.io/ look decent

File: agent/agent.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	agentMetrics, err := buildAgentMetrics()
+	if err != nil {
+		log.Println("buildAgentMetrics error", err)
+		return
+	}
+
+	currentTime := getCurrentTime()
+	agentConfig, err := a.ApiService.Checkin(currentTime, a.LocalAgent, a.PushAgent, a.Ephemeral, agentMetrics)
+	if err != nil {
+		log.Println("Checkin error", err)
```
Comment (moribellamy) [inline agent/agent.go:46]: not sure how golang logging works, is there a way to escalate this so it seems more urgent than, for example, "fetching configuration" above? maybe log to a different channel?

File: agent/metrics.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			meter := metric.Snapshot()
+			addRate(stats, name, meter.Count(), meter.Rate1(), meter.RateMean())
+		}
+	})
+	return json.Marshal(stats)
+}
+
+func addHisto(stats map[string]interface{}, name string, min, max int64, mean float64, percentiles []float64) {
+	// convert from nanoseconds to milliseconds
+	stats[combine(name, "duration.min")] = float64(min) / 1000000
```
Comment (moribellamy) [inline agent/metrics.go]: 1e6 might be more fun and less prone to accidental zeroes

File: api/api.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	AgentID   string
+	Hostname  string
+	Token     string
+	Version   string
+}
+
+func (service *WavefrontAPIService) GetConfig(currentMillis, bytesLeft, bytesPerMinute, currentQueueSize int64) (*config.AgentConfig, error) {
+	log.Println("GetConfig")
+
+	apiURL := service.ServerURL + getConfigSuffix
```
Comment (moribellamy) [inline api/api.go:41]: found this on stackoverflow:

u, err := url.Parse("http://foo")
u.Path = path.Join(u.Path, "bar.html")
s := u.String() // prints http://foo/bar.html

File: points/forwarder.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		})
+	}
+	log.Printf("%s: exiting flushPoints", f.name)
+}
+
+func (f *DefaultPointForwarder) stop() {
+	f.pushTicker.Stop()
+}
+
+func min(x, y int) int {
```
Comment (moribellamy) [inline points/forwarder.go:61]: math.min or something?

File: points/forwarder.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	currLen := len(f.points)
+	batchSize := min(currLen, f.maxFlushSize)
+	batchPoints := f.points[:batchSize]
+	f.points = f.points[batchSize:currLen]
+	f.mtx.Unlock()
+	return batchPoints
+}
+
+func (f *DefaultPointForwarder) buffer(points []string) {
+	f.mtx.Lock()
```
Comment (moribellamy) [inline points/forwarder.go:79]: is "defer unlock" idiomatic here? or maybe golang has something similar to python's context guard?

File: points/parser/element.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+	for tok != EOF && tok > literal_beg && tok < literal_end {
+		buf.WriteString(lit)
+		tok, lit = p.scan()
+	}
+	p.unscan()
+	return buf.String(), nil
+}
+
+func getCurrentTime() int64 {
+	return time.Now().UnixNano() / 1000000000
```
Comment (moribellamy) [inline points/parser/element.go]: 1e9 maybe

**Review by moribellamy — COMMENTED** [review]
Nice work. Seems like a fun project.

I just have some superficial comments, I defer to your judgement on all of them.

File: points/forwarder.go (hunk 5 above)
Comment (vikramraman) [reply points/forwarder.go:61]: Strangely, the in-built Math.min (https://golang.org/pkg/math/#Min) only takes floats. It is apparently idiomatic in Go to just write a simple min function for other cases.

File: points/forwarder.go (hunk 6 above)
Comment (vikramraman) [reply points/forwarder.go:79]: Yes, makes sense to do `defer f.mtx.Unlock()` at the top here.

File: Makefile (hunk 1 above)
Comment (vikramraman) [reply Makefile:1]: Thanks for sharing! Will keep this in mind as this project evolves.

File: agent/metrics.go (hunk 3 above)
Comment (vikramraman) [reply agent/metrics.go]: Makes sense. Changed to 1e6 and 1e9 (comment below).

File: agent/agent.go (hunk 2 above)
Comment (vikramraman) [reply agent/agent.go:46]: The default Go log package doesn't support log levels. Most projects appear to use third party pkgs or roll out their own logging framework. We may want to do that down the line.

It is possible to have multiple loggers writing to different output files. But this seems suffice for now.

File: api/api.go (hunk 4 above)
Comment (vikramraman) [reply api/api.go:41]: Very useful! Will include this in the near future as part of a broader change to cache some of the URLs that are static for every call (GetConfig, Checkin and AgentConfigProcessed).

**Comment by panghy** [comment]
I'll lean on @moribellamy's comments and pull this in for now

**Pull request merged**